### PR TITLE
Validate monitoring username at parse time

### PR DESCRIPTION
**Pull request by danhermann**
Provides parse-time validation for `AUTH_USERNAME_SETTING` as described in #47711.

### CODE REVIEW

File: x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/exporter/http/HttpExporter.java [hunk 1]
```diff
@@ -177,7 +177,46 @@ public void validate(final List<String> hosts, final Map<Setting<?>, Object> set
      */
     public static final Setting.AffixSetting<String> AUTH_USERNAME_SETTING =
             Setting.affixKeySetting("xpack.monitoring.exporters.","auth.username",
-                    (key) -> Setting.simpleString(key, Property.Dynamic, Property.NodeScope, Property.Filtered));
+                    (key) -> Setting.simpleString(
+                        key,
+                        new Setting.Validator<String>() {
+                            @Override
+                            public void validate(String password) {
+
```
Comment (jakelandis) [inline x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/exporter/http/HttpExporter.java]: nit: comment to why empty

**Review by jakelandis — APPROVED** [review]
LGTM (one nit)

**Comment by danhermann** [comment]
@elastic/es-core-infra, this work covers both monitoring and settings if you want to review it.

**Comment by danhermann** [comment]
@elasticmachine, run elasticsearch-ci/2, elasticsearch-ci/bwc

**Comment by danhermann** [comment]
@elasticmachine, run elasticsearch-ci/2

**Comment by danhermann** [comment]
@elasticmachine, run elasticsearch-ci/bwc

**Comment by danhermann** [comment]
@elasticmachine, run elasticsearch-ci/2

**Comment by danhermann** [comment]
@elasticmachine, run elasticsearch-ci/bwc

**Comment by danhermann** [comment]
@elasticmachine, run elasticsearch-ci/2

**Comment by danhermann** [comment]
@elasticmachine, run elasticsearch-ci/bwc

**Comment by danhermann** [comment]
@elasticmachine update branch

File: x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/exporter/http/HttpExporter.java [hunk 2]
```diff
@@ -177,7 +177,46 @@ public void validate(final List<String> hosts, final Map<Setting<?>, Object> set
...
+                                    HttpExporter.AUTH_USERNAME_SETTING.getNamespace(
+                                        HttpExporter.AUTH_USERNAME_SETTING.getConcreteSetting(key));
+                                final String password =
+                                    (String) settings.get(AUTH_PASSWORD_SETTING.getConcreteSettingForNamespace(namespace));
+
+                                // password must be specified along with username for any auth
+                                if (Strings.isNullOrEmpty(username) == false) {
+                                    if (Strings.isNullOrEmpty(password)) {
+                                        throw new SettingsException(
+                                            "[" + AUTH_USERNAME_SETTING.getConcreteSettingForNamespace(namespace).getKey() + "] without [" +
```
Comment (rjernst) [inline x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/exporter/http/HttpExporter.java]: Could we have a more explicit message? This is something a user sees that must be acted upon. Perhaps something for the form "XXX is set but YYY is missing"

**Review by rjernst — APPROVED** [review]
LGTM, one minor comment

**Comment by danhermann** [comment]
@elasticmachine run elasticsearch-ci/bwc

**Comment by danhermann** [comment]
@elasticmachine run elasticsearch-ci/default-distro

**Comment by danhermann** [comment]
@elasticmachine run elasticsearch-ci/bwc

**Comment by danhermann** [comment]
@elasticmachine update branch

**Comment by danhermann** [comment]
@elasticmachine update branch

**Comment by danhermann** [comment]
@elasticmachine run elasticsearch-ci/2

**Pull request merged**